### PR TITLE
Big refactoring

### PR DESCRIPTION
**Pull request by samuelgarcia**
@alejoe91 @magland @colehurwitz @mhhennig 

Hi all.
here the very very begiing of the refactoring of spikeinterface:
  * make a unique package with submodule not imported
  * transpose trace (times, channels)
  * make cache at base level (no need CacheExtractor)
  * annotation/properties/features handle at base level and in a vector way. easy to dump/cache)
  * handle multiple segment at Base level So ned two class:
    * Recording list of RecordingSegment
    * RecordingSegment: trace for one segment
    * Sorting is a list of SortingSegment
    * SortingSegment : handle spike for one segment.
  * will use probeinterface for handling probe (so all read_prb is no more here) 

The "core" submodule will contain more or less the spikeextractor main files but not reader in subfolder.
The "extractor" submodule will contain all extractors but most of then will be neo based, I hope.

This is a total draft. I prefer to show now before it become too big.

I would be happy to have discussion/feedback.

### CODE REVIEW

File: spikeinterface/core/mytypes.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+class Order(Enum):
+    """Enum for memory order for raw traces in signal block
+    """
+    C = 1  # C order
+    F = 2  # Fortran order
+    K = 3  # Original order
+
+
+class SampleIndex(int):
```
Comment (alejoe91) [inline spikeinterface/core/mytypes.py:11]: @samuelgarcia what would you need this for? 

I would avoid to over engineer the classes and objects used..

File: spikeinterface/extractors/neoextractors/mearecextractor.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            if locations.shape[1] == 3:
+                locations = locations[:, 1:]
+        
+        print(locations.shape)
+        print(self._main_ids)
+        self.set_channel_locations(locations)
+    
+    @staticmethod
+    def write_recording(recording, save_path, check_suffix=True):
+        # Alessio : I think we don't need this
```
Comment (alejoe91) [inline spikeinterface/extractors/neoextractors/mearecextractor.py:50]: I agree, remove plz

File: spikeinterface/sorters/basesorter.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        This should not be used anymore.
+        """
+        # setup recording
+        self._setup_recording(self.recording, self.output_folder)
+        
+        # compute
+        self.compute_from_folder(self.output_folder, raise_error=True)
+
+    @staticmethod
+    def compute_from_folder(output_folder, raise_error=True):
```
Comment (alejoe91) [inline spikeinterface/sorters/basesorter.py:138]: I'd call it `run_from_folder`

File: spikeinterface/sorters/basesorter.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+            else:
+                print(f'{sorter_name} run time {run_time:0.2f}s')
+
+        if has_error and raise_error:
+            raise SpikeSortingError(f"Spike sorting failed. You can inspect the runtime trace in "
+                                    f"the {sorter_name}.log of the output folder.'")
+        
+        return run_time
+
+    def get_result(self):
```
Comment (alejoe91) [inline spikeinterface/sorters/basesorter.py:207]: note that the `raise_error` need to be propagated here too: see https://github.com/SpikeInterface/spikesorters/pull/207

**Comment by samuelgarcia** [comment]
I will do a first in this branch.

@alejoe91 : please no direct commit into this branch!!!!!
Only PR with own repo.

**Review by alejoe91 — APPROVED** [review]

**Pull request merged**